### ISSUE TITLE
Release v0.17.0

### ISSUE THREAD

**Issue by TobiGr** [TeamNewPipe/NewPipe#2483]
#### Playback resume #2288
- Resume streams where you stopped last time
#### Downloader Enhancements #2149 
- Use Storage Access Framework to store downloads on external SD-cards
- New mp4 muxer
- Optionally change the download directory before starting a download
- Respect metered networks 

---------

### Improved
- Removed gema strings #2295
- Handle (auto)rotation changes during activity lifecycle #2444
- Make long-press menus consistent #2368

---------

### Fixed
- Fixed selected subtitle track name not being shown #2394
- Do not crash when check for app update fails (GitHub version) #2423 
- Fixed downloads stuck at 99.9% #2440
- Update play queue metadata #2453 
- [SoundCloud] Fixed crash when loading playlists TeamNewPipe/NewPipeExtractor#170
- [YouTube] Fixed duration can not be paresd TeamNewPipe/NewPipeExtractor#177
- [YouTube] Fixed crash on videos which are age restricted TeamNewPipe/NewPipeExtractor#177

---------

Corresponding PR: #2482
Corresponding PR for the extractor: https://github.com/TeamNewPipe/NewPipeExtractor/pull/179

### APK for testing and finding bugs

NewPipe-v0.17.0_RC4:
- [release](https://github.com/TeamNewPipe/NewPipe/files/3488761/NewPipe-v0.17.0_release_RC4.apk.zip)
- [debug](https://github.com/TeamNewPipe/NewPipe/files/3488763/NewPipe_v0.17.0_debug_RC4.apk.zip)

**Please report all new bugs below.**

**Comment by theScrabi** [comment]
I'd also open an issue for this, as I think people read more issues than pull requests.

Also remember to merge weblate:
- [ ] Weblate has been merged

**Comment by theScrabi** [comment]
Found an issue, the history does not show up directly after you opened the history view, and if you do anything it will immediately disappear again.

**Comment by TobiGr** [comment]
> Also remember to merge weblate:

Done.

> Found an issue, the history does not show up directly after you opened the history view, and if you do anything it will immediately disappear again.

Yes it takes quite way to long to load. But on my device nothing disappears.

**Comment by theScrabi** [comment]
I think its a little weird that not even the loading cycle is shown during that time.

**Comment by tusch001** [comment]
export settings database crashes due to missing permission although write access to storage in general is allowed (Android 6). Maybe also use SAF like when asking for the audio / video download folder?

error message (sorry can't append a file):
{
   "user_action": "ui error",
   "request": "",
   "content_language": "GB",
   "service": "none",
   "package": "org.schabi.newpipe",
   "version": "0.17.0",
   "os": "Linux Android 6.0.1 - 23",
   "time": "2019-08-01 07:32",
   "exceptions": [
      "java.io.FileNotFoundException: \/storage\/8900-7EE7\/NewPipeData-20190801_093211.zip: open failed: EACCES (Permission denied)\n\tat libcore.io.IoBridge.open(IoBridge.java:452)\n\tat java.io.FileOutputStream.<init>(FileOutputStream.java:87)\n\tat java.io.FileOutputStream.<init>(FileOutputStream.java:127)\n\tat java.io.FileOutputStream.<init>(FileOutputStream.java:116)\n\tat org.schabi.newpipe.settings.ContentSettingsFragment.exportDatabase(ContentSettingsFragment.java:154)\n\tat org.schabi.newpipe.settings.ContentSettingsFragment.onActivityResult(ContentSettingsFragment.java:139)\n\tat android.support.v4.app.FragmentActivity.onActivityResult(FragmentActivity.java:160)\n\tat android.app.Activity.dispatchActivityResult(Activity.java:6497)\n\tat android.app.ActivityThread.deliverResults(ActivityThread.java:3695)\n\tat android.app.ActivityThread.handleSendResult(ActivityThread.java:3742)\n\tat android.app.ActivityThread.access$1400(ActivityThread.java:150)\n\tat android.app.ActivityThread$H.handleMessage(ActivityThread.java:1393)\n\tat android.os.Handler.dispatchMessage(Handler.java:102)\n\tat android.os.Looper.loop(Looper.java:148)\n\tat android.app.ActivityThread.main(ActivityThread.java:5417)\n\tat java.lang.reflect.Method.invoke(Native Method)\n\tat com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)\n\tat com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)\nCaused by: android.system.ErrnoException: open failed: EACCES (Permission denied)\n\tat libcore.io.Posix.open(Native Method)\n\tat libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)\n\tat libcore.io.IoBridge.open(IoBridge.java:438)\n\t... 17 more\n"
   ],
   "user_comment": ""
}

**Comment by phatloot123** [comment]
I've found an error unfortunately I can't reproduce it :/. I was opening a youtube link in another app with newpipe -> show info and then i went into to the settings and when i went back to the video i got this error:

## Exception
* __User Action:__ ui error
* __Request:__ App crash, UI failure
* __Content Language:__ GB
* __Service:__ none
* __Version:__ 0.17.0
* __OS:__ Linux Android 8.0.0 - 26


<details><summary><b>Crash log</b></summary><p>

```
java.lang.RuntimeException: Unable to resume activity {org.schabi.newpipe/org.schabi.newpipe.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.List org.schabi.newpipe.extractor.stream.StreamInfo.getVideoStreams()' on a null object reference
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3790)
	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3830)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1746)
	at android.os.Handler.dispatchMessage(Handler.java:105)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6944)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.List org.schabi.newpipe.extractor.stream.StreamInfo.getVideoStreams()' on a null object reference
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.setupActionBar(VideoDetailFragment.java:659)
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.onResume(VideoDetailFragment.java:250)
	at android.support.v4.app.Fragment.performResume(Fragment.java:2498)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1501)
	at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
	at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
	at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3241)
	at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:223)
	at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:538)
	at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:527)
	at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:172)
	at android.app.Activity.performResume(Activity.java:7385)
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3765)
	... 8 more
java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.List org.schabi.newpipe.extractor.stream.StreamInfo.getVideoStreams()' on a null object reference
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.setupActionBar(VideoDetailFragment.java:659)
	at org.schabi.newpipe.fragments.detail.VideoDetailFragment.onResume(VideoDetailFragment.java:250)
	at android.support.v4.app.Fragment.performResume(Fragment.java:2498)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1501)
	at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
	at android.support.v4.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
	at android.support.v4.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:3241)
	at android.support.v4.app.FragmentController.dispatchResume(FragmentController.java:223)
	at android.support.v4.app.FragmentActivity.onResumeFragments(FragmentActivity.java:538)
	at android.support.v4.app.FragmentActivity.onPostResume(FragmentActivity.java:527)
	at android.support.v7.app.AppCompatActivity.onPostResume(AppCompatActivity.java:172)
	at android.app.Activity.performResume(Activity.java:7385)
	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3765)
	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3830)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1746)
	at android.os.Handler.dispatchMessage(Handler.java:105)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6944)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)

```
</p></details>
<hr>

**Comment by theScrabi** [comment]
I found out why my chonnic is so slow: I have a huge amout of data in there. So its underperforming dramatically. However it used to work well before the update, so there must be something new that has bad runtime. One should run the profiler on this.

**Comment by theScrabi** [comment]
Found another bug:
Importing settings from an exported database file is not getting applied.
As you can see in the video, I've set the default resolution to 720p, and exported the database, and on import I said "yes" when NP asked me to restore the settings, however the 720p did not get restored, which also applied to all other settings:

![bla mp4](https://user-images.githubusercontent.com/1891273/62344687-56536480-b4ef-11e9-99bd-d414ed36e5c8.gif)
![Download video](https://github.com/TeamNewPipe/NewPipe/files/3459630/bla.m4v.zip)

**Comment by Gymcap** [comment]
First, this update looks great. All these fixes and improvements are very exciting. 

Second, is there any way [#2309](https://github.com/TeamNewPipe/NewPipe/pull/2309) could be implemented as an experimental setting?

**Comment by kapodamy** [comment]
In "what is new" if the list is loaded quicky (previuosly browing all channels on the feed). The empty list message is displayed
![Screenshot_1564852824](https://user-images.githubusercontent.com/16857233/62415015-c3532100-b5f9-11e9-94fb-0dca5b9d6e96.png)

**Comment by yausername** [comment]
> In "whats new" if the list is loaded quicky (previuosly browing all channels on the feed). The empty list message is displayed
> ![Screenshot_1564852824](https://user-images.githubusercontent.com/16857233/62415015-c3532100-b5f9-11e9-94fb-0dca5b9d6e96.png)

Related issues - 
- https://github.com/TeamNewPipe/NewPipe/pull/2203#pullrequestreview-264540283
- some comments are hidden under video thumbnail/detail. Can only see by scrolling up.

I managed to fix it with https://github.com/yausername/NewPipe/commit/985f5f718051d20e74dd893c4390ff8367671f37 Not sure of any side effects though.
@nv95

**Comment by tusch001** [comment]
What's new list: around the sixth entry a video is always shown twice (only one).

**Comment by theScrabi** [comment]
I found another issue: The player starts a video from the beginning when you switch away from it, to another activity, and then return.

![clip mp4](https://user-images.githubusercontent.com/1891273/62542140-7d959300-b85b-11e9-829e-f3daabb25b56.gif)

**Comment by yausername** [comment]
some nit-picking. Is this watched duration thing really required? The red bar seems good enough to me.
![playback-resume](https://user-images.githubusercontent.com/5203007/62587007-a9595d00-b8dd-11e9-814f-d61dadec405c.png)

**Comment by theScrabi** [comment]
I thought thats cool :D so you know exaclty the time where you are.

**Comment by theScrabi** [comment]
@tusch001 what you postet is not a regression and should therefore not be fixed now.

**Comment by theScrabi** [comment]
For me the only big issue now is that thing that when you return to the player the possition is lost. Can someone confirm that?

When thats done it consider release. @TobiGr  ? What do u think?

**Comment by TobiGr** [comment]
@theScrabi I agree. Everything else can be fixed in a future version.

**Comment by thegodofplay** [comment]
I am not able to clear playback position history. is there any option to do it. And also after turning off this feature the indicator doesn't show in list but when I click on the video it show up.And also video starts from saved position even feature is off.
I tried clearing history with cache(in app info too) but it doesn't clear up.

is anyone having same problem.

newpipe   : 17.0 R3
Android    : 8.1
Language: en

**Comment by thegodofplay** [comment]
@theScrabi. I am not having the problem you mentioned about playback from start. I tried same video and is resume from right point.

Maybe you are using a custom rom.
If you are using custom rom:
Some custom roms has very aggressive ram management system so it can influence running process as you go background in order to save power or Maybe rom isn't stable.

**Comment by TobiGr** [comment]
> I am not able to clear playback position history. is there any option to do it.

The positions is only deleted when deleting the history.

@theScrabi I am not able to reproduce this on Android, too. Only with Lineage 15.1. Which Lineage version are you using?

**Comment by yausername** [comment]
@TobiGr @theScrabi 
check PR https://github.com/TeamNewPipe/NewPipe/pull/2496. I think this should be part of this release as it is a regression bug.

**Comment by thegodofplay** [comment]
My bad. After turning off playback position the progress bar doesn't show up anywhere, its working fine. But it still logging positions even if feature is off and start showing previous positions(of time when feature is off) when you enable it again. And i mentioned above that i cleaed history but it doesn't work. Here is the video: 
https://streamable.com/olcvg

**Comment by dhnasa** [comment]
> I found another issue: The player starts a video from the beginning when you switch away from it, to another activity, and then return.

@theScrabi I can reproduce this issue on Lineage 16.0

**Comment by theScrabi** [comment]
@TobiGr I have Lineage 16.0 aka Android 9.
More exact: `16.0-20190729-microG-bardock`

@thegodofplay I know that some images might have problems, os do all Android phones have. But we try to make it work as well as possible. However Lineage is not one of the least used Roms.

**Comment by TobiGr** [comment]
> But it still logging positions even if feature is off and start showing previous positions(of time when feature is off) when you enable it again

@thegodofplay This should not happen and is IMO a new blocker @theScrabi

**Comment by TobiGr** [comment]
@thegodofplay Fixed the history bug in https://github.com/TeamNewPipe/NewPipe/commit/3cd485069d19b681a29faefa8f1fda148a5976e8

**Comment by thegodofplay** [comment]
I want ask that is it intentional or a issue that both the video and  audio download paths are empty by default so newpipe ask for paths everytime until user set one by himself in the settings.

**Comment by theScrabi** [comment]
I noticed that to but it happend not always

**Comment by TobiGr** [comment]
I was not able to fix the problem mentioned in https://github.com/TeamNewPipe/NewPipe/issues/2483#issuecomment-518660305
@nv95, @theScrabi, @yausername or @Stypox: Can one of you  take a look at it please?

**Comment by theScrabi** [comment]
Awesome. :D Thx @TobiGr . I will build later. I gat quite some busy day yesterday.

**Comment by justanidea** [comment]
I found another nasty bug: the local playlists content duplicates again and again and this is so bothering
How to reproduce:
1- open a playlist
2- long press the first element and click "enqueue in background"
3- skip the song from the notification
4- you are done ! You have all the songs twice. And i realized it just after skipping 3 songs so my playlists went from 15 songs to 45. But for a reason i still couldnt determine, this doesnt happen all the time, and if you leave the playlist and come back, everything come back as before - but sometimes not.

**Comment by theScrabi** [comment]
Uploaded RC4.

**Comment by theScrabi** [comment]
@TobiGr the bug from https://github.com/TeamNewPipe/NewPipe/issues/2483#issuecomment-518660305 seems still to be there in RC4.

@justanidea I could not reproduce hits behaviour. Also did you test this with 0.16.4? If the bug exists there it's not a regression and should therefore not be fixed now.

**Comment by TobiGr** [comment]
@theScrabi that's why I asked for help

**Comment by theScrabi** [comment]
I see, I missread sorry. I was already taking a look into the code but could not find the cause yet as well.

**Comment by yausername** [comment]
@theScrabi @TobiGr @Stypox https://github.com/TeamNewPipe/NewPipe/issues/2483#issuecomment-518660305 seems to be caused by https://github.com/TeamNewPipe/NewPipe/pull/2482/commits/7c9ef58acd0971ef47c78f4d3247d388991044bc

It is working fine if I revert that commit (https://github.com/TeamNewPipe/NewPipe/pull/2482/commits/7c9ef58acd0971ef47c78f4d3247d388991044bc). Although I can't completely understand the reason behind this commit.

**Comment by theScrabi** [comment]
How about we release now, because 0.17.0 is less broken then 0.16.4. We could release now and fix on the masterbranch later

**Comment by TobiGr** [comment]
Sounds good!

**Comment by theScrabi** [comment]
All right

**Comment by Pentaphon** [comment]
@theScrabi are you releasing today?

**Comment by theScrabi** [comment]
No I'm sick today. @TheAssassin you have the keys could you take care of it, otherwise I will tomorrow.

**Comment by Stypox** [comment]
@yausername That commit fixed [this problem](https://github.com/TeamNewPipe/NewPipe/pull/2368#issuecomment-514994273). But what I did was just a workaround. I will now try harder and push directly to `release_v0.17.0` (as it seems like I have the permission ;-))

**Comment by goxr3plus** [comment]
Release it babieeeesssssss

**Comment by Stypox** [comment]
I can't reproduce https://github.com/TeamNewPipe/NewPipe/issues/2483#issuecomment-518660305 on RC4
@yausername @theScrabi

**Comment by Stypox** [comment]
@yausername @theScrabi @nv95 I could not reproduce the issue, but I tried to fix it anyway. I'm now disposing `stateLoader` when the player is destroyed, and probably this is the correct way to do it. Check https://github.com/Stypox/NewPipe/commit/5191907af0f5abfcb45b6dcb09cc56a8924ce5db

Here is the APK for testing: 
[app-debug_1.zip](https://github.com/TeamNewPipe/NewPipe/files/3490402/app-debug_1.zip)

**Comment by theScrabi** [comment]
@Stypox you fixed it :D

**Comment by yausername** [comment]
> I could not reproduce the issue

It was happening on android P.

**Comment by theScrabi** [comment]
Well done. Thank you very much, it has been a pleasure to work with all of you. :bowing_man: 


![](https://media.giphy.com/media/NnGGHE0muVqpO/giphy.gif) ![](https://media.giphy.com/media/l4q8cJzGdR9J8w3hS/giphy.gif)

**Comment by justanidea** [comment]
Actually the playlist bug doesnt happen anymore, i guess it was linked to the older version of Np 0.17.0, before all these fixes

**Comment by theScrabi** [comment]
I had to pull back 0.17.0 due to issues with the documents app on android 5.0, and the slow performance when having old databases that would need to be imported into 0.17.0.
